File: design_patterns/C++/Structural/composite.cxx
#include <iostream>
#include <vector>
#include <string>

class Employee {
  public:
	virtual std::string getName() = 0;
	virtual void setSalary(float salary) = 0;
	virtual float getSalary() = 0;
	virtual std::vector<std::string> getRoles() = 0;
};

class Developer: public Employee {
  protected:
	float salary = 0.0;
	std::string name = "";
	std::vector<std::string> roles;

  public:
	Developer(std::string name, float salary) {
		this->name = name;
		this->salary = salary;
	}

	std::string getName() {
		return this->name;
	}

	void setSalary(float salary) {
		this->salary = salary;
	}

	float getSalary() {
		return this->salary;
	}

	std::vector<std::string> getRoles() {
		return this->roles;
	}
};

class Designer: public Employee {
  protected:
	std::string name = "";
	float salary = 0.0;
	std::vector<std::string> roles;

  public:
	Designer(std::string name, float salary) {
		this->name = name;
		this->salary = salary;
	}

	std::string getName() {
		return this->name;
	}

	void setSalary(float salary) {
		this->salary = salary;
	}

	float getSalary() {
		return this->salary;
	}

	std::vector<std::string> getRoles() {
		return this->roles;
	}
};

class Organization {
  protected:
	std::vector<Employee*> employees;

  public:
	void addEmployee(Employee* employee) {
		this->employees.push_back(employee);
	}

	float getNetSalaries() {
		auto netSalaries = 0.0;
		for (auto employee : this->employees) {
			netSalaries += employee->getSalary();
		}
		return netSalaries;
	}
};

int main() {
	auto john = new Developer("John Doe", 12000);
	auto jane = new Designer("Jane Doe", 10000);
	auto organization = new Organization();
	organization->addEmployee(john);
	organization->addEmployee(jane);
	std::cout << "Net salaries: " << organization->getNetSalaries() << std::endl;
	// ===================
	// Net salaries: 22000
	// ===================
	return 0;
}

File: design_patterns/Python/Structural/composite.py
class Employee:

    def get_name(self):
        raise NotImplementedError

    def set_salary(self, salary):
        raise NotImplementedError

    def get_salary(self):
        raise NotImplementedError

    def get_roles(self):
        raise NotImplementedError


class Developer(Employee):

    def __init__(self, name, salary):
        self.name = name
        self.salary = salary

    def get_name(self):
        return self.name

    def set_salary(self, salary):
        self.salary = salary

    def get_salary(self):
        return self.salary

    def get_roles(self):
        return self.roles


class Designer(Employee):

    def __init__(self, name, salary):
        self.name = name
        self.salary = salary

    def get_name(self):
        return self.name

    def set_salary(self, salary):
        self.salary = salary

    def get_salary(self):
        return self.salary

    def get_roles(self):
        return self.roles


class Organization:

    def __init__(self):
        self.employees = []

    def add_employee(self, employee):
        self.employees.append(employee)

    def get_net_salaries(self):
        net_salary = 0
        for employee in self.employees:
            net_salary += employee.get_salary()
        return net_salary


if __name__ == '__main__':

    john = Developer('John Doe', 12000)
    jane = Designer('Jane Doe', 10000)
    organization = Organization()
    organization.add_employee(john)
    organization.add_employee(jane)
    assert organization.get_net_salaries() == 22000
